File: src/main/java/io/github/ozkanpakdil/grepwise/service/AuthService.java
package io.github.ozkanpakdil.grepwise.service;

import io.github.ozkanpakdil.grepwise.grpc.*;
import io.github.ozkanpakdil.grepwise.model.Role;
import io.github.ozkanpakdil.grepwise.model.User;
import io.github.ozkanpakdil.grepwise.repository.RoleRepository;
import io.github.ozkanpakdil.grepwise.repository.UserRepository;
import io.grpc.stub.StreamObserver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Service implementation for handling authentication operations.
 */
@Service
public class AuthService extends AuthServiceGrpc.AuthServiceImplBase {
    private static final Logger logger = LoggerFactory.getLogger(AuthService.class);

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final TokenService tokenService;
    private final AuditLogService auditLogService;
    private final BCryptPasswordEncoder passwordEncoder;

    public AuthService(UserRepository userRepository, RoleRepository roleRepository, 
                      TokenService tokenService, AuditLogService auditLogService) {
        this.userRepository = userRepository;
        this.roleRepository = roleRepository;
        this.tokenService = tokenService;
        this.auditLogService = auditLogService;
        this.passwordEncoder = new BCryptPasswordEncoder();
        
        // Initialize default roles
        roleRepository.initializeDefaultRoles();
        
        // Create a default admin user if no users exist
        if (userRepository.count() == 0) {
            User admin = new User();
            admin.setUsername("admin");
            admin.setEmail("admin@grepwise.io");
            admin.setPassword(passwordEncoder.encode("admin"));
            admin.setFirstName("Admin");
            admin.setLastName("User");
            
            // Add ADMIN role
            Role adminRole = roleRepository.findByName("ADMIN");
            if (adminRole != null) {
                admin.addRole(adminRole);
            }
            
            userRepository.save(admin);
        }
    }

    @Override
    public void register(RegisterRequest request, StreamObserver<AuthResponse> responseObserver) {
        logger.info("Registration attempt for username: {}", request.getUsername());
        
        // Check if username or email already exists
        if (userRepository.existsByUsername(request.getUsername())) {
            logger.warn("Registration failed: Username already exists: {}", request.getUsername());
            
            // Log the failed registration attempt
            auditLogService.createAuthAuditLog(
                "REGISTER", 
                "FAILURE", 
                request.getUsername(), 
                "Registration failed: Username already exists"
            );
            
            responseObserver.onNext(AuthResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("Username already exists")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        if (userRepository.existsByEmail(request.getEmail())) {
            logger.warn("Registration failed: Email already exists: {}", request.getEmail());
            
            // Log the failed registration attempt
            auditLogService.createAuthAuditLog(
                "REGISTER", 
                "FAILURE", 
                request.getUsername(), 
                "Registration failed: Email already exists"
            );
            
            responseObserver.onNext(AuthResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("Email already exists")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Create new user
        User user = new User();
        user.setUsername(request.getUsername());
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        user.setFirstName(request.getFirstName());
        user.setLastName(request.getLastName());
        
        // Add USER role
        Role userRole = roleRepository.findByName("USER");
        if (userRole != null) {
            user.addRole(userRole);
        }
        
        user = userRepository.save(user);
        
        // Generate tokens
        String accessToken = tokenService.generateToken(user);
        String refreshToken = tokenService.generateRefreshToken(user);
        
        // Log the successful registration
        logger.info("User registered successfully: {}", user.getUsername());
        auditLogService.createAuthAuditLog(
            "REGISTER", 
            "SUCCESS", 
            user.getUsername(), 
            "User registered successfully"
        );
        
        // Build response
        AuthResponse response = AuthResponse.newBuilder()
                .setSuccess(true)
                .setMessage("User registered successfully")
                .setAccessToken(accessToken)
                .setRefreshToken(refreshToken)
                .setExpiresAt(tokenService.getExpirationDateFromToken(accessToken).getTime())
                .setUser(convertToGrpcUser(user))
                .build();
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void login(LoginRequest request, StreamObserver<AuthResponse> responseObserver) {
        logger.info("Login attempt for username: {}", request.getUsername());
        
        // Find user by username
        User user = userRepository.findByUsername(request.getUsername());
        
        // Check if user exists and password matches
        if (user == null || !passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            logger.warn("Login failed: Invalid username or password for: {}", request.getUsername());
            
            // Log the failed login attempt
            auditLogService.createAuthAuditLog(
                "LOGIN", 
                "FAILURE", 
                request.getUsername(), 
                "Login failed: Invalid username or password"
            );
            
            responseObserver.onNext(AuthResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("Invalid username or password")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Generate tokens
        String accessToken = tokenService.generateToken(user);
        String refreshToken = tokenService.generateRefreshToken(user);
        
        // Log the successful login
        logger.info("User logged in successfully: {}", user.getUsername());
        auditLogService.createAuthAuditLog(
            "LOGIN", 
            "SUCCESS", 
            user.getUsername(), 
            "Login successful"
        );
        
        // Build response
        AuthResponse response = AuthResponse.newBuilder()
                .setSuccess(true)
                .setMessage("Login successful")
                .setAccessToken(accessToken)
                .setRefreshToken(refreshToken)
                .setExpiresAt(tokenService.getExpirationDateFromToken(accessToken).getTime())
                .setUser(convertToGrpcUser(user))
                .build();
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void validateToken(ValidateTokenRequest request, StreamObserver<ValidateTokenResponse> responseObserver) {
        String token = request.getToken();
        
        // Extract username from token for logging (might be null if token is invalid)
        String username = null;
        try {
            username = tokenService.getUsernameFromToken(token);
            logger.debug("Token validation attempt for user: {}", username);
        } catch (Exception e) {
            logger.debug("Token validation attempt with potentially invalid token");
        }
        
        // Validate token
        if (!tokenService.validateToken(token)) {
            logger.debug("Token validation failed: Invalid token");
            
            // Log the failed token validation (only if we could extract a username)
            if (username != null) {
                auditLogService.createAuthAuditLog(
                    "TOKEN_VALIDATE", 
                    "FAILURE", 
                    username, 
                    "Token validation failed: Invalid token"
                );
            }
            
            responseObserver.onNext(ValidateTokenResponse.newBuilder()
                    .setValid(false)
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Get user ID and roles from token
        String userId = tokenService.getUserIdFromToken(token);
        List<String> roles = tokenService.getRolesFromToken(token);
        long expiresAt = tokenService.getExpirationDateFromToken(token).getTime();
        
        // Log successful token validation (using debug level since this happens frequently)
        logger.debug("Token validated successfully for user: {}", username);
        
        // We don't log successful token validations to avoid excessive audit logs
        // since this operation happens frequently
        
        // Build response
        ValidateTokenResponse response = ValidateTokenResponse.newBuilder()
                .setValid(true)
                .setUserId(userId)
                .addAllRoles(roles)
                .setExpiresAt(expiresAt)
                .build();
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void refreshToken(RefreshTokenRequest request, StreamObserver<AuthResponse> responseObserver) {
        String refreshToken = request.getRefreshToken();
        
        // Extract username from token for logging (might be null if token is invalid)
        String username = null;
        try {
            username = tokenService.getUsernameFromToken(refreshToken);
            logger.info("Token refresh attempt for user: {}", username);
        } catch (Exception e) {
            logger.info("Token refresh attempt with invalid token");
        }
        
        // Validate refresh token
        if (!tokenService.validateToken(refreshToken)) {
            logger.warn("Token refresh failed: Invalid refresh token");
            
            // Log the failed token refresh
            auditLogService.createAuthAuditLog(
                "TOKEN_REFRESH", 
                "FAILURE", 
                username != null ? username : "unknown", 
                "Token refresh failed: Invalid refresh token"
            );
            
            responseObserver.onNext(AuthResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("Invalid refresh token")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Get user ID from token
        String userId = tokenService.getUserIdFromToken(refreshToken);
        User user = userRepository.findById(userId);
        
        if (user == null) {
            logger.warn("Token refresh failed: User not found for ID: {}", userId);
            
            // Log the failed token refresh
            auditLogService.createAuthAuditLog(
                "TOKEN_REFRESH", 
                "FAILURE", 
                username != null ? username : "unknown", 
                "Token refresh failed: User not found"
            );
            
            responseObserver.onNext(AuthResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("User not found")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Generate new access token
        String accessToken = tokenService.generateToken(user);
        
        // Log the successful token refresh
        logger.info("Token refreshed successfully for user: {}", user.getUsername());
        auditLogService.createAuthAuditLog(
            "TOKEN_REFRESH", 
            "SUCCESS", 
            user.getUsername(), 
            "Token refreshed successfully"
        );
        
        // Build response
        AuthResponse response = AuthResponse.newBuilder()
                .setSuccess(true)
                .setMessage("Token refreshed successfully")
                .setAccessToken(accessToken)
                .setRefreshToken(refreshToken)
                .setExpiresAt(tokenService.getExpirationDateFromToken(accessToken).getTime())
                .setUser(convertToGrpcUser(user))
                .build();
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void logout(LogoutRequest request, StreamObserver<LogoutResponse> responseObserver) {
        logger.info("Logout request received");
        
        // In a real implementation, we would invalidate the tokens
        // For now, we just return a success response and log the event
        
        // Log the logout event
        // Since we don't have user information in the logout request,
        // we'll log it with an "unknown" username
        auditLogService.createAuthAuditLog(
            "LOGOUT", 
            "SUCCESS", 
            "unknown", 
            "User logged out successfully"
        );
        
        LogoutResponse response = LogoutResponse.newBuilder()
                .setSuccess(true)
                .setMessage("Logout successful")
                .build();
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void getCurrentUser(GetCurrentUserRequest request, StreamObserver<io.github.ozkanpakdil.grepwise.grpc.User> responseObserver) {
        String token = request.getAccessToken();
        
        // Validate token
        if (!tokenService.validateToken(token)) {
            // Return an empty user if token is invalid
            responseObserver.onNext(io.github.ozkanpakdil.grepwise.grpc.User.newBuilder().build());
            responseObserver.onCompleted();
            return;
        }
        
        // Get user ID from token
        String userId = tokenService.getUserIdFromToken(token);
        User user = userRepository.findById(userId);
        
        if (user == null) {
            // Return an empty user if user not found
            responseObserver.onNext(io.github.ozkanpakdil.grepwise.grpc.User.newBuilder().build());
            responseObserver.onCompleted();
            return;
        }
        
        // Build response
        io.github.ozkanpakdil.grepwise.grpc.User response = convertToGrpcUser(user);
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void updateUser(UpdateUserRequest request, StreamObserver<io.github.ozkanpakdil.grepwise.grpc.User> responseObserver) {
        String token = request.getAccessToken();
        
        // Validate token
        if (!tokenService.validateToken(token)) {
            // Return an empty user if token is invalid
            responseObserver.onNext(io.github.ozkanpakdil.grepwise.grpc.User.newBuilder().build());
            responseObserver.onCompleted();
            return;
        }
        
        // Get user ID from token
        String userId = tokenService.getUserIdFromToken(token);
        User user = userRepository.findById(userId);
        
        if (user == null) {
            // Return an empty user if user not found
            responseObserver.onNext(io.github.ozkanpakdil.grepwise.grpc.User.newBuilder().build());
            responseObserver.onCompleted();
            return;
        }
        
        // Update user information
        if (!request.getEmail().isEmpty()) {
            user.setEmail(request.getEmail());
        }
        
        if (!request.getFirstName().isEmpty()) {
            user.setFirstName(request.getFirstName());
        }
        
        if (!request.getLastName().isEmpty()) {
            user.setLastName(request.getLastName());
        }
        
        user = userRepository.save(user);
        
        // Build response
        io.github.ozkanpakdil.grepwise.grpc.User response = convertToGrpcUser(user);
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    @Override
    public void changePassword(ChangePasswordRequest request, StreamObserver<ChangePasswordResponse> responseObserver) {
        String token = request.getAccessToken();
        
        // Validate token
        if (!tokenService.validateToken(token)) {
            responseObserver.onNext(ChangePasswordResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("Invalid token")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Get user ID from token
        String userId = tokenService.getUserIdFromToken(token);
        User user = userRepository.findById(userId);
        
        if (user == null) {
            responseObserver.onNext(ChangePasswordResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("User not found")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Check if current password matches
        if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPassword())) {
            responseObserver.onNext(ChangePasswordResponse.newBuilder()
                    .setSuccess(false)
                    .setMessage("Current password is incorrect")
                    .build());
            responseObserver.onCompleted();
            return;
        }
        
        // Update password
        user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        userRepository.save(user);
        
        // Build response
        ChangePasswordResponse response = ChangePasswordResponse.newBuilder()
                .setSuccess(true)
                .setMessage("Password changed successfully")
                .build();
        
        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    /**
     * Convert a model User to a gRPC User.
     *
     * @param modelUser The model User to convert
     * @return The converted gRPC User
     */
    private io.github.ozkanpakdil.grepwise.grpc.User convertToGrpcUser(User modelUser) {
        return io.github.ozkanpakdil.grepwise.grpc.User.newBuilder()
                .setId(modelUser.getId())
                .setUsername(modelUser.getUsername())
                .setEmail(modelUser.getEmail())
                .setFirstName(modelUser.getFirstName())
                .setLastName(modelUser.getLastName())
                .addAllRoles(modelUser.getRoleNames())
                .setCreatedAt(modelUser.getCreatedAt())
                .setUpdatedAt(modelUser.getUpdatedAt())
                .setEnabled(modelUser.isEnabled())
                .build();
    }
}

File: frontend/src/services/AuthService.ts
import { AuthServiceClient } from '../generated/AuthServiceServiceClientPb';
import {
  RegisterRequest,
  LoginRequest,
  AuthResponse,
  ValidateTokenRequest,
  ValidateTokenResponse,
  RefreshTokenRequest,
  LogoutRequest,
  LogoutResponse,
  GetCurrentUserRequest,
  User,
  UpdateUserRequest,
  ChangePasswordRequest,
  ChangePasswordResponse
} from '../generated/authservice_pb';

/**
 * Service for interacting with authentication operations via gRPC-Web.
 */
class AuthService {
  private client: AuthServiceClient;
  private accessToken: string | null = null;
  private refreshToken: string | null = null;
  private user: User | null = null;

  constructor() {
    // Create a client instance using the gRPC-Web client
    // The URL should match the gRPC server endpoint with the gRPC-Web proxy
    this.client = new AuthServiceClient('http://localhost:9090');
    
    // Load tokens from localStorage if available
    this.loadTokens();
  }

  /**
   * Load tokens from localStorage.
   */
  private loadTokens(): void {
    try {
      this.accessToken = localStorage.getItem('accessToken');
      this.refreshToken = localStorage.getItem('refreshToken');
    } catch (error) {
      console.error('Error loading tokens from localStorage:', error);
    }
  }

  /**
   * Save tokens to localStorage.
   * 
   * @param accessToken Access token
   * @param refreshToken Refresh token
   */
  private saveTokens(accessToken: string, refreshToken: string): void {
    try {
      localStorage.setItem('accessToken', accessToken);
      localStorage.setItem('refreshToken', refreshToken);
      this.accessToken = accessToken;
      this.refreshToken = refreshToken;
    } catch (error) {
      console.error('Error saving tokens to localStorage:', error);
    }
  }

  /**
   * Clear tokens from localStorage.
   */
  private clearTokens(): void {
    try {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('refreshToken');
      this.accessToken = null;
      this.refreshToken = null;
      this.user = null;
    } catch (error) {
      console.error('Error clearing tokens from localStorage:', error);
    }
  }

  /**
   * Register a new user.
   * 
   * @param username Username
   * @param email Email
   * @param password Password
   * @param firstName First name
   * @param lastName Last name
   * @returns Promise with auth response
   */
  async register(
    username: string,
    email: string,
    password: string,
    firstName: string,
    lastName: string
  ): Promise<AuthResponse> {
    const request = new RegisterRequest();
    request.setUsername(username);
    request.setEmail(email);
    request.setPassword(password);
    request.setFirstName(firstName);
    request.setLastName(lastName);
    
    return new Promise((resolve, reject) => {
      this.client.register(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          if (response.getSuccess()) {
            this.saveTokens(response.getAccessToken(), response.getRefreshToken());
            this.user = response.getUser();
          }
          resolve(response);
        }
      });
    });
  }

  /**
   * Login with username and password.
   * 
   * @param username Username
   * @param password Password
   * @returns Promise with auth response
   */
  async login(username: string, password: string): Promise<AuthResponse> {
    const request = new LoginRequest();
    request.setUsername(username);
    request.setPassword(password);
    
    return new Promise((resolve, reject) => {
      this.client.login(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          if (response.getSuccess()) {
            this.saveTokens(response.getAccessToken(), response.getRefreshToken());
            this.user = response.getUser();
          }
          resolve(response);
        }
      });
    });
  }

  /**
   * Validate a token.
   * 
   * @param token Token to validate
   * @returns Promise with validation response
   */
  async validateToken(token: string = this.accessToken || ''): Promise<ValidateTokenResponse> {
    const request = new ValidateTokenRequest();
    request.setToken(token);
    
    return new Promise((resolve, reject) => {
      this.client.validateToken(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Refresh an access token using a refresh token.
   * 
   * @param refreshToken Refresh token (uses stored token if not provided)
   * @returns Promise with auth response
   */
  async refreshToken(refreshToken: string = this.refreshToken || ''): Promise<AuthResponse> {
    const request = new RefreshTokenRequest();
    request.setRefreshToken(refreshToken);
    
    return new Promise((resolve, reject) => {
      this.client.refreshToken(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          if (response.getSuccess()) {
            this.saveTokens(response.getAccessToken(), response.getRefreshToken());
            this.user = response.getUser();
          }
          resolve(response);
        }
      });
    });
  }

  /**
   * Logout and invalidate tokens.
   * 
   * @returns Promise with logout response
   */
  async logout(): Promise<LogoutResponse> {
    const request = new LogoutRequest();
    if (this.accessToken) {
      request.setAccessToken(this.accessToken);
    }
    if (this.refreshToken) {
      request.setRefreshToken(this.refreshToken);
    }
    
    return new Promise((resolve, reject) => {
      this.client.logout(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          this.clearTokens();
          resolve(response);
        }
      });
    });
  }

  /**
   * Get current user information.
   * 
   * @returns Promise with user information
   */
  async getCurrentUser(): Promise<User> {
    if (!this.accessToken) {
      return Promise.reject(new Error('No access token available'));
    }
    
    const request = new GetCurrentUserRequest();
    request.setAccessToken(this.accessToken);
    
    return new Promise((resolve, reject) => {
      this.client.getCurrentUser(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          this.user = response;
          resolve(response);
        }
      });
    });
  }

  /**
   * Update user information.
   * 
   * @param email Email
   * @param firstName First name
   * @param lastName Last name
   * @returns Promise with updated user
   */
  async updateUser(
    email: string,
    firstName: string,
    lastName: string
  ): Promise<User> {
    if (!this.accessToken) {
      return Promise.reject(new Error('No access token available'));
    }
    
    const request = new UpdateUserRequest();
    request.setAccessToken(this.accessToken);
    request.setEmail(email);
    request.setFirstName(firstName);
    request.setLastName(lastName);
    
    return new Promise((resolve, reject) => {
      this.client.updateUser(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          this.user = response;
          resolve(response);
        }
      });
    });
  }

  /**
   * Change password.
   * 
   * @param currentPassword Current password
   * @param newPassword New password
   * @returns Promise with change password response
   */
  async changePassword(
    currentPassword: string,
    newPassword: string
  ): Promise<ChangePasswordResponse> {
    if (!this.accessToken) {
      return Promise.reject(new Error('No access token available'));
    }
    
    const request = new ChangePasswordRequest();
    request.setAccessToken(this.accessToken);
    request.setCurrentPassword(currentPassword);
    request.setNewPassword(newPassword);
    
    return new Promise((resolve, reject) => {
      this.client.changePassword(request, null, (err, response) => {
        if (err) {
          reject(err);
        } else {
          resolve(response);
        }
      });
    });
  }

  /**
   * Check if the user is authenticated.
   * 
   * @returns True if the user is authenticated
   */
  isAuthenticated(): boolean {
    return !!this.accessToken;
  }

  /**
   * Get the current user.
   * 
   * @returns Current user or null
   */
  getUser(): User | null {
    return this.user;
  }

  /**
   * Get the access token.
   * 
   * @returns Access token or null
   */
  getAccessToken(): string | null {
    return this.accessToken;
  }
}

// Export a singleton instance
export const authService = new AuthService();
export default AuthService;